### PR TITLE
C#: Data flow performance improvements

### PR DESCRIPTION
**Pull request by hvitved**
This PR has a [quite positive impact on performance](https://git.semmle.com/gist/tom/4112fd995620cd05826d3bab4ddce928) (internal link), as the default LGTM suite contains many data-flow-based queries.

Commit-by-commit review is encouraged.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll [hunk 1]
```diff
@@ -1100,7 +1125,10 @@ module DataFlow {
           config.isSource(node) or
           jumpStep(_, node) or
           node instanceof ParameterNode or
-          node instanceof OutNode
+          node instanceof OutNode or
+          node instanceof NormalReturnNode or
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll]: Just `node instanceof ReturnNode` ?

File: csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll [hunk 2]
```diff
@@ -1111,9 +1139,18 @@ module DataFlow {
...
-          node instanceof ArgumentNode or
-          node instanceof ReturnNode or
+          jumpStep(node, _)
+          or
+          node instanceof ArgumentNode
+          or
+          node instanceof ReturnNode
+          or
+          exists(ReturnNode rn | localFlowStep(node, rn, config) |
+            rn instanceof NormalReturnNode or
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll]: I don't see how the body of this `exists` can be false? So perhaps
```
localFlowStep(node, any(ReturnNode rn), config)
```

File: csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll [hunk 3]
```diff
@@ -591,18 +581,47 @@ module DataFlow {
...
-          c.(Callable).canYieldReturn(e) or
-          callableReturnsOutOrRef(c, _, e)
-        )
-      }
+    class NormalReturnNode extends ReturnNode, TNormalReturnNode {
+      override Type getType() { result = this.getEnclosingCallable().getReturnType() }
+
+      override DotNet::Callable getEnclosingCallable() { this = TNormalReturnNode(result) }
+
+      override string toString() { result = "return" }
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll]: `"return " + getEnclosingCallable().getName()`
Could that be useful?

File: csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll [hunk 4]
```diff
@@ -591,18 +581,47 @@ module DataFlow {
     abstract private class ReturnNode extends Node { }
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll]: Could it be useful to have `getAReturnedExpr()` on this class?

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/dataflow/DataFlow.qll]: It is an internal class, so I think we should only add it if we need it later on.

**Comment by hvitved** [comment]
> Wait for dist-compare results.

The [new report](https://git.semmle.com/gist/tom/36d3a2eb9b0fa9f4d2ff07e9c53831cc) is ready (internal link), and things still look good.

**Pull request merged**